### ISSUE TITLE
When switching scale factor in OSD on recent MacOS builds it switches twice on one keypress

### ISSUE THREAD

**Issue by sdsnatcher73** [openMSX/openMSX#1169]
Reproduce:
1. Open openMSX (e.g. build 339 from openmsx.dev) on MacOS (10.14.5 in my case). 
2. Open the OSD menu.
3. Navigate to Video Settings and open submenu.
4. Navigate to Scale Factor.
5. Change Scale Factor to 1 by pressing left cursor repeatedly.
6. Press right cursor once shortly, observe scale factor first switches to 2 then later to 3.

**Comment by MBilderbeek** [comment]
I can't reproduce it on Linux. Can you reproduce it on Windows?

**Comment by sdsnatcher73** [comment]
No does not happen on Windows again just MacOS.

**Comment by niekvlessert** [comment]
Don't have the issue on 10.13.6. I'm running the latest build from openmsx.dev, which is 330, not 339? Could be related to the keyboard issue from https://github.com/openMSX/openMSX/issues/1170?

**Comment by sdsnatcher73** [comment]
I meant 330 build indeed (typo). Here is my 330 build. It may indeed be related to #1170.
If you need me to try something just let me know...
[openmsx-0.15.0-330-g8cc1c31d5-mac-x86_64-bin.dmg.zip](https://github.com/openMSX/openMSX/files/3319317/openmsx-0.15.0-330-g8cc1c31d5-mac-x86_64-bin.dmg.zip)

**Comment by niekvlessert** [comment]
Tried your build; to complicate things; I don't have this OSD issue, but I do have the laggy keyboard issue.

**Comment by sdsnatcher73** [comment]
At least it clarifies that this issue is dependent on where openMSX is running both your build and mine are fine on 10.13.6 but have issues on 10.14.5.

**Comment by sdsnatcher73** [comment]
This issue still exists after #1170 was fixed...

**Comment by MBilderbeek** [comment]
It's strange that no one can reproduce it, even not @niekvlessert on his Mac. Can you try again, Niek?
It's very hard to solve an issue that we cannot reproduce. There should be nothing special about that case. Looks like it gets 2 events or something? Can you see something when you enable that logging patch I made for the other ticket?

**Comment by sdsnatcher73** [comment]
The patch does not help figure this out. Every key press when in the OSD leads to this log:

Got KEYDOWN, but not handling yet...
SDL_KEYDOWN at handleEvent
Created KEYDOWN event and send to triggerOsdControlEvents

But some keypresses lead to a single change while another leads to 2 changes. From looking at the code after triggerOsdControlEvents the event gets passed around quite a bit.. It may be in the time it spends in that chain that causes the event to be processed twice on occasion. (Actually the changing of the scale of course takes quite some time, the OSD key release may come late because of that).

**Comment by niekvlessert** [comment]
Tried it again on the 337 build, no issues at all. You tried High Sierra as well right..? Then it might be hardware related in some way... Tried 337 on El Capitan on my older Macbook Pro, doesn't have the issue. I used the build created by my usual environment on openmsx.dev. @peterhanraets, you're sure you can reproduce it with that right?

**Comment by MBilderbeek** [comment]
@peterhanraets Does it matter whether you're using the SDL or SDLGL-PP renderer? (The switching speed may make a difference somehow.)

**Comment by sdsnatcher73** [comment]
I added some extra debug output
[InputEventGenerator.cc.gz](https://github.com/openMSX/openMSX/files/3383423/InputEventGenerator.cc.gz)

This shows the following output:
Got KEYDOWN, but not handling yet...
SDL_KEYDOWN at handleEvent
Created KEYDOWN event and send to triggerOsdControlEvents
Received K_RIGHT in triggerOsdControlEventsFromKeyEvent
Now in osdControlChangeButtom
Now in setNewOsdControlButtonState
Distributing OsdControlPressEvent event
Received K_RIGHT in triggerOsdControlEventsFromKeyEvent
Now in osdControlChangeButtom
Now in setNewOsdControlButtonState
Distributing OsdControlReleaseEvent event

This does appear to be alright, where are the OsdControlPressEvent and OsdControlReleaseEvent events handled?

**Comment by sdsnatcher73** [comment]
> @peterhanraets Does it matter whether you're using the SDL or SDLGL-PP renderer? (The switching speed may make a difference somehow.)

Yes! it **does not** happen with SDL renderer but it **does** happen with SDLGL-PP!

**Comment by sdsnatcher73** [comment]
> Tried it again on the 337 build, no issues at all. You tried High Sierra as well right..? Then it might be hardware related in some way... Tried 337 on El Capitan on my older Macbook Pro, doesn't have the issue. I used the build created by my usual environment on openmsx.dev. @peterhanraets, you're sure you can reproduce it with that right?

Yes I can reproduce it with your build. Both on Mojave and High Sierra. I am using a (relatively) slow MacBook Retina 2012 (the 12" model) with a 1.1Ghz Core m3 processor.

**Comment by niekvlessert** [comment]
Mine is a 2011 Macbook Pro with a I7-2635QM. It uses Intel HD Graphics 3000. What is the exact type of the GPU in the m3?

**Comment by MBilderbeek** [comment]
@peterhanraets Try to add logging to print which events are sent by SDL, in InputEventGeneratir, inside the while loop of poll(). Perhaps you get the same event twice, after the scale_factor switch?

**Comment by sdsnatcher73** [comment]
> Mine is a 2011 Macbook Pro with a I7-2635QM. It uses Intel HD Graphics 3000. What is the exact type of the GPU in the m3?

Intel HD Graphics 5300 1536 MB

**Comment by sdsnatcher73** [comment]
> @peterhanraets Try to add logging to print which events are sent by SDL, in InputEventGeneratir, inside the while loop of poll(). Perhaps you get the same event twice, after the scale_factor switch?

The SDL key is only received once. My impression is that it is related to how an event is translated into a state (which is again polled possibly)?

So this is how I see the flow now (but part is guessing)
1. SDL_KEYDOWN is received
2. KEYDOWN is handled
3. event is transferred to triggerOsdControlEventsFromKeyEvent
4. OsdControlButtonState is changed
5. OsdControlPressEvent is distributed
6. Somehow scale factor is changed by OSD menu (TCL?) If this is due to state (and not PressEvent) that might be causing the issue.
7. event is again transfered to triggerOsdControlEventsFromKeyEvent
8. OsdControlButtonState is changed
9. OsdControlReleaseEvent is distributed

Now if the OsdControlReleaseEvent is delayed and the OsdControlButtonState is polled again it may trigger the second scale change...

**Comment by sdsnatcher73** [comment]
I added a debug print in createSurface, this is what is logged now when the key press is handled twice:
Got KEYDOWN, but not handling yet...
SDL_KEYDOWN at handleEvent
Created KEYDOWN event and send to triggerOsdControlEvents
Received K_LEFT in triggerOsdControlEventsFromKeyEvent
Now in osdControlChangeButtom
Now in setNewOsdControlButtonState
Distributing OsdControlPressEvent event
Now in createSurface...
Received K_LEFT in triggerOsdControlEventsFromKeyEvent
Now in osdControlChangeButtom
Now in setNewOsdControlButtonState
Distributing OsdControlReleaseEvent event
**Now in createSurface...**

**Comment by sdsnatcher73** [comment]
So it seems _osd_menu.tcl increases or decreases scale_factor by 1 on LEFT/RIGHT. How is that picked up by the core or video subsystem??

**Comment by MBilderbeek** [comment]
The OSD menu code just changes a setting. And settings are listened to by the code that should respond to them.

There is no polling involved in the code (except for polling events at SDL), everything is event based.

Perhaps it's easier if you join us on IRC to discuss directly? :)

**Comment by grauw** [comment]
I tried to reproduce but I can not.

Tried my own 51bad0652 build, as well as builds 330 and 337 from openmsx.dev. Using XCode 11.3.1 and MacOS Catalina 10.15.1 on a MacBook Pro (16-inch, 2019). I don’t have my mid-2012 Macbook Pro Retina anymore unfortunately (that one had the 650M GPU).

**Comment by sdsnatcher73** [comment]
Just build 635 from scratch on macOS 10.15.2 with Xcode 11.3.1 and this is still happening on my (admittedly relatively low specced) MacBook Retina with a core m3 processor.

**Comment by sdsnatcher73** [comment]
When trying the SDL specific scanline issue, I noticed this issue does not happen when SDL renderer is active.

**Comment by MBilderbeek** [comment]
@grauw Can you double check on SDLGL-PP?

**Comment by grauw** [comment]
I was testing on SDLGL-PP already since it was mentioned earlier.

Maybe I got the rhythm wrong or something? I press left a bunch of times and then press right once to see if it switches twice. I tried different timings, but it kept switching as expected.

**Comment by sdsnatcher73** [comment]
No that is how I do it.Somehow this is only an issue (apparantly) on my core m3 macBook. I could imagine it may be an issue on slower Linux or Windows systems as well but I don't have any to test it on.

It may be circumvented by changing it to a sub menu where you see the 4 options, select one and only when enter is pressed the setting is changed (which is then picked up).

**Comment by sdsnatcher73** [comment]
This is now happening on my Surface Laptop Go running Windows 11 with openMSX 17.0-355

**Comment by MBilderbeek** [comment]
is this issue still relevant on the latest development build?

**Comment by MBilderbeek** [comment]
@sdsnatcher73 is it still relevant in openMSX 20.0? If yes, please reopen.
